go/store/nbs/table_reader.go

```diff
@@ -764,141 +746,126 @@ func (tr tableReader) readAtOffsetsWithCB(
+		return tr.readCompressedAtOffsets(ctx, rb, found, stats)
 	})
 }
 
 func (tr tableReader) getManyAtOffsets(
 	ctx context.Context,
-	reqs []getRecord,
+	eg *errgroup.Group,
 	offsetRecords offsetRecSlice,
-	foundChunks chan<- *chunks.Chunk,
-	wg *sync.WaitGroup,
-	ae *atomicerr.AtomicError,
+	found func(*chunks.Chunk),
 	stats *Stats,
-) {
-	tr.getManyAtOffsetsWithReadFunc(ctx, reqs, offsetRecords, wg, ae, stats, func(
+) error {
+	return tr.getManyAtOffsetsWithReadFunc(ctx, eg, offsetRecords, stats, func(
 		ctx context.Context,
-		readStart, readEnd uint64,
-		reqs []getRecord,
-		offsets offsetRecSlice,
+		rb readBatch,
 		stats *Stats) error {
-
-		return tr.readAtOffsets(ctx, readStart, readEnd, reqs, offsets, foundChunks, stats)
+		return tr.readAtOffsets(ctx, rb, found, stats)
 	})
 }
 
+type readBatch offsetRecSlice
+
+func (r readBatch) Start() uint64 {
+	return r[0].offset
+}
+
+func (r readBatch) End() uint64 {
+	last := r[len(r)-1]
+	return last.offset + uint64(last.length)
+}
+
+func (s readBatch) ExtractChunkFromRead(buff []byte, idx int) (CompressedChunk, error) {
+	rec := s[idx]
+	chunkStart := rec.offset - s.Start()
+	return NewCompressedChunk(hash.Hash(*rec.a), buff[chunkStart:chunkStart+uint64(rec.length)])
+}
+
+func toReadBatches(offsets offsetRecSlice, blockSize uint64) []readBatch {
+	res := make([]readBatch, 0)
+	var batch readBatch
+	for i := 0; i < len(offsets); {
+		rec := offsets[i]
+		if batch == nil {
+			batch = readBatch{rec}
+			i++
+			continue
+		}
+
+		if _, canRead := canReadAhead(rec, batch.End(), blockSize); canRead {
+			batch = append(batch, rec)
+			i++
+			continue
+		}
+
+		res = append(res, batch)
+		batch = nil
+	}
+	if batch != nil {
+		res = append(res, batch)
+	}
+	return res
+}
+
 func (tr tableReader) getManyAtOffsetsWithReadFunc(
 	ctx context.Context,
-	reqs []getRecord,
+	eg *errgroup.Group,
 	offsetRecords offsetRecSlice,
-	wg *sync.WaitGroup,
-	ae *atomicerr.AtomicError,
 	stats *Stats,
 	readAtOffsets func(
 		ctx context.Context,
-		readStart, readEnd uint64,
-		reqs []getRecord,
-		offsets offsetRecSlice,
+		rb readBatch,
 		stats *Stats) error,
-) {
-	type readBatch struct {
-		batch     offsetRecSlice
-		readStart uint64
-		readEnd   uint64
-	}
-	buildBatches := func(batchCh chan<- readBatch) {
-		// |offsetRecords| contains all locations within the table
-		// which must be search in sorted order and without
-		// duplicates. Now scan forward, grouping sequences of reads
-		// into larger physical reads.
-
-		var batch offsetRecSlice
-		var readStart, readEnd uint64
-
-		for i := 0; i < len(offsetRecords); {
-			if ae.IsSet() {
-				break
-			}
-
-			rec := offsetRecords[i]
-			length := rec.length
-
-			if batch == nil {
-				batch = make(offsetRecSlice, 1)
-				batch[0] = offsetRecords[i]
-				readStart = rec.offset
-				readEnd = readStart + uint64(length)
-				i++
-				continue
+) error {
+	batches := toReadBatches(offsetRecords, tr.blockSize)
```

Comment:
Calculating these before spawning goroutines did not seem like a big loss and was a decent simplification. Creating the little `readBatch` helper type seemed like a win.